src/nexusrpc/handler/_core.py

```diff
@@ -0,0 +1,523 @@
+)
+from ._serializer import LazyValue
+from ._types import MISSING_TYPE, InputT, OutputT, ServiceHandlerT
+
+
+class UnknownServiceError(RuntimeError):
+    """Raised when a request contains a service name that does not match a service handler."""
+
+    pass
+
+
+class UnknownOperationError(RuntimeError):
+    """Raised when a request contains an operation name that does not match an operation handler."""
+
+    pass
+
+
+@dataclass
+class Handler:
+    """
+    A Nexus handler, managing a collection of Nexus service handlers.
+
+    Operation requests are delegated to a :py:class:`ServiceHandler` based on the service
+    name in the operation context.
+    """
+
+    service_handlers: dict[str, ServiceHandler]
+
+    def __init__(
+        self,
+        user_service_handlers: Sequence[Any],
+        executor: Optional[SyncExecutor] = None,
+    ):
+        """Initialize a :py:class:`Handler` instance from user service handler instances.
+
+        The user service handler instances must have been decorated with the
+        :py:func:`@nexusrpc.handler.service_handler` decorator.
+
+        Args:
+            user_service_handlers: A sequence of user service handlers.
+            executor: An executor to run non-`async def` operation handlers in.
+        """
+        self.executor = executor
+        self.service_handlers = {}
+        for sh in user_service_handlers:
+            if isinstance(sh, type):
+                raise TypeError(
+                    f"Expected a service instance, but got a class: {type(sh)}. "
+                    "Nexus service handlers must be supplied as instances, not classes."
+                )
+            # Users may register ServiceHandler instances directly.
+            if not isinstance(sh, ServiceHandler):
+                # It must be a user service handler instance (i.e. an instance of a class
+                # decorated with @nexusrpc.handler.service_handler).
+                sh = ServiceHandler.from_user_instance(sh)
+            if sh.service.name in self.service_handlers:
+                raise RuntimeError(
+                    f"Service '{sh.service.name}' has already been registered."
+                )
+            if self.executor is None:
+                for op_name, operation_handler in sh.operation_handlers.items():
+                    if not is_async_callable(operation_handler.start):
+                        raise RuntimeError(
+                            f"Service '{sh.service.name}' operation '{op_name}' start must be an `async def` if no executor is provided."
+                        )
+                    if not is_async_callable(operation_handler.cancel):
+                        raise RuntimeError(
+                            f"Service '{sh.service.name}' operation '{op_name}' cancel must be an `async def` if no executor is provided."
+                        )
+            self.service_handlers[sh.service.name] = sh
+
+    async def start_operation(
```

Comment:
So to use Nexus handlers, users _must_ use async? There is no concept of an all-sync Nexus?

Reply:
The PR now has `def` versions of `Handler`, `SyncOperationHandler`, and `sync_operation` in the `nexusrpc.handler.syncio` module.

I think that given that, we should not support `consume_sync` on `LazyValue` but instead, either

(A) have `nexusrpc.syncio.LazyValue`

(B) create a parallel tree of everything under `nexusrpc.syncio`. This can contain handler components and `LazyValue` initially, and later on client and server.

(B) sounds best to me.